File: 384. Shuffle an Array/C++/Solution.cpp
class Solution {
public:
    vector<int> origin;
    vector<int> nums;

    Solution(vector<int> nums) {
        origin = nums;
        this->nums = nums;
    }

    /** Resets the array to its original configuration and return it. */
    vector<int> reset() {
        return origin;
    }

    /** Returns a random shuffling of the array. */
    vector<int> shuffle() {
        int length = nums.size();
        for (int i = 0; i < length; i++) {
            int swapIndex = rand() % length;
            int temp = nums[i];
            nums[i] = nums[swapIndex];
            nums[swapIndex] = temp;
        }
        return nums;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * vector<int> param_1 = obj.reset();
 * vector<int> param_2 = obj.shuffle();
 */

File: 384. Shuffle an Array/Python/Solution.py
class Solution(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.origin = nums
        self.nums = nums[:]

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        :rtype: List[int]
        """
        return self.origin

    def shuffle(self):
        """
        Returns a random shuffling of the array.
        :rtype: List[int]
        """
        nums = self.nums
        length = len(nums)
        for i in range(length):
            swap_index = random.randrange(length)
            nums[i], nums[swap_index] = nums[swap_index], nums[i]
        return self.nums


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()
